### PR TITLE
allow waiting for enough peers in gossipsub

### LINKED ISSUE #454: support waiting until a gossipsub topic has enough peers
Right now, most tests that build a gossipsub mesh use sleeps to wait for the mesh to build:

https://github.com/libp2p/go-libp2p-pubsub/blob/7ef0669764b94ed11e8b5b0128ee408942ebe5eb/gossipsub_test.go#L62-L63

The indexer-reference-provider codebase has the same problem:

https://github.com/filecoin-project/indexer-reference-provider/blob/3ec1151ca682a437c6e38727aa00b1087e5a0276/core/engine/engine_test.go#L186-L188

It seems like it would be enough to wait for enough peers to be connected before publishing advertisements.
Right now, the following options are available:

1) Sleep. This is what all the tests do, but one has to use a very long sleep to reduce the chances of flakes. Even with multi-second sleeps, the CI over at indexer-reference-provider still ran into failures at times.

2) Poll `PubSubRouter.EnoughPeers` until it returns true. Would work, but just like any other form of polling, it's a tradeoff between slowness and overhead.

3) Poll `Topic.ListPeers` until its length reaches a minimum. Same polling problems as 1.

4) Use `WithReadiness` when publishing. Unfortunately, this option only works when `WithDiscovery` is used, and the indexer doesn't use that. The option seems to be a no-op in those cases.

I think we should support some way to do this without sleeping nor polling, nor without requiring `WithDiscovery`. Here are two proposed API additions:

5) Modify `WithReadiness` so that it works in all configurations. Without `WithDiscovery`, it would simply make `Topic.Publish` block until enough peers are connected (or until Publish's `ctx` is cancelled).

6) Add another method, such as `func (*Topic) WaitEnoughPeers(context.Context, int) bool`, which would block until enough peers are connected or ctx is cancelled.

5 seems preferrable to 6, for the sake of making existing APIs more consistently useful and not adding redundant APIs. However, I'm not familiar enough with `WithReadiness` to tell if it's the best approach internally.

### PR DESCRIPTION
**Pull request by mvdan**
(see commit message)

### CODE REVIEW

**Comment by mvdan** [comment]
FYI @adlrocha @masih @willscott for the time.Sleep TODOs in go-legs and indexer-reference-provider.

**Comment by mvdan** [comment]
@willscott another question is how to expose this via go-legs. Right now it calls NewGossipSub, which creates a GossipSubRouter and feeds it to NewPubSub. Unfortunately, that is then hidden behind an unexported field in `PubSub.rt`, so neither go-legs nor its downstreams can access it to call methods like WaitEnoughPeers.

One option would be to add another method, like:

    func (*PubSub) Router() PubSubRouter

Thoughts, @vyzo?

**Comment by mvdan** [comment]
Yet another option is to leave it up to the downstream to poll `EnoughPeers` in a loop, since that's already part of the `PubSubRouter` interface. That would certainly work, but then we'd be stuck with polling and sleeping forever, without the ability to do better like the TODO says. So I think supporting a standard method like "wait for enough peers" is worthwhile.

I'm new to this module and its design, though, so maybe there's a better design to be had here.

**Review by vyzo — CHANGES_REQUESTED** [review]
This should be part of the pubsub interface, not the router; all routers implement enough peers.
Also, thread safety...

**Comment by vyzo** [comment]
Actually it should be in the topic interface.

**Comment by vyzo** [comment]
and there is also machinery for topic related events, which can make implementation of this much nicer.

**Comment by vyzo** [comment]
and there is already an option that more or less does what you want, it's called `WithReadiness`.

**Comment by mvdan** [comment]
> and there is already an option that more or less does what you want, it's called `WithReadiness`.

I saw that, but it doesn't do anything at all in our case, because we're not using `WithDiscovery`.
Maybe we could teach that option to also work without `WithDiscovery`? It's unclear to me what API would be nicest.

> Actually it should be in the topic interface.

That's fine by me; I was just following `EnoughPeers`. I can rewrite this PR if we can agree on what API to add or modify.

**Comment by vyzo** [comment]
then open an issue to discuss and agree on an api.

**Comment by mvdan** [comment]
> then open an issue to discuss and agree on an api.

Done: https://github.com/libp2p/go-libp2p-pubsub/issues/454

**Comment by vyzo** [comment]
thanks.

**Comment by mvdan** [comment]
I've updated the code to no longer be racy, I think - the added `WithReadiness` logic inside `Publish` now uses the `eval` channel to call the `ready` func, which then calls `EnoughPeers`. Can you take another look?

File: gossipsub.go
Comment (vyzo) [inline gossipsub.go]: can we remove this?

File: topic.go
Comment (vyzo) [inline topic.go]: we should use a ticker here.

**Review by vyzo — COMMENTED** [review]
i would say this is good enough.

However, please don't change the gossipsub tests to remove the sleep, we want to test mesh formation with heartbeats.

**Comment by mvdan** [comment]
Hm, I'm not sure I understand removing the sleeps from tests. Sleeping for two seconds or sleep-polling until the equivalent condition is met seems basically the same, just that sleeping makes the tests slower. I'm happy to revert those changes though, as I'm not the maintainer.

If I'm not touching any of the existing tests, then it sounds like I need to add a new test to cover this feature.

**Comment by vyzo** [comment]
it's not the same condition; one allows the mesh to form, the other waits
for number of peers.

Yes, please add a new test for the readiness condition.

File: topic.go
Comment (mvdan) [reply topic.go]: done

**Comment by mvdan** [comment]
All done, I think.

File: topic.go
Comment (vyzo) [inline topic.go]: an improvement here is to only instantiate the ticker after making an optimistic attempt; this way we only pay the price once.

**Review by vyzo — COMMENTED** [review]
looks pretty good - we just need to avoid unnecessary ticker allocations once we've got enough peers.

File: topic.go
Comment (mvdan) [reply topic.go]: done

**Review by vyzo — APPROVED** [review]
thank you!

**Pull request merged**
